cpp/src/copying/contiguous_split.cu

```diff
@@ -813,282 +907,303 @@ struct size_of_helper {
-  src_buf_info* d_src_buf_info = reinterpret_cast<src_buf_info*>(
-    reinterpret_cast<uint8_t*>(d_indices_and_source_info.data()) + indices_size);
-  size_type* d_offset_stack =
-    reinterpret_cast<size_type*>(reinterpret_cast<uint8_t*>(d_indices_and_source_info.data()) +
-                                 indices_size + src_buf_info_size);
-
-  // compute splits -> indices.
-  h_indices[0]              = 0;
-  h_indices[num_partitions] = input.column(0).size();
-  std::copy(splits.begin(), splits.end(), std::next(h_indices));
-
-  // setup source buf info
-  setup_source_buf_info(input.begin(), input.end(), h_src_buf_info, h_src_buf_info, stream);
-
-  // HtoD indices and source buf info to device
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    d_indices, h_indices, indices_size + src_buf_info_size, cudaMemcpyDefault, stream.value()));
-
-  // packed block of memory 2. partition buffer sizes and dst_buf_info structs
-  std::size_t const buf_sizes_size =
-    cudf::util::round_up_safe(num_partitions * sizeof(std::size_t), split_align);
-  std::size_t const dst_buf_info_size =
-    cudf::util::round_up_safe(num_bufs * sizeof(dst_buf_info), split_align);
-  // host-side
-  std::vector<uint8_t> h_buf_sizes_and_dst_info(buf_sizes_size + dst_buf_info_size);
-  std::size_t* h_buf_sizes = reinterpret_cast<std::size_t*>(h_buf_sizes_and_dst_info.data());
-  dst_buf_info* h_dst_buf_info =
-    reinterpret_cast<dst_buf_info*>(h_buf_sizes_and_dst_info.data() + buf_sizes_size);
-  // device-side
-  rmm::device_buffer d_buf_sizes_and_dst_info(
-    buf_sizes_size + dst_buf_info_size, stream, rmm::mr::get_current_device_resource());
-  std::size_t* d_buf_sizes     = reinterpret_cast<std::size_t*>(d_buf_sizes_and_dst_info.data());
-  dst_buf_info* d_dst_buf_info = reinterpret_cast<dst_buf_info*>(
-    static_cast<uint8_t*>(d_buf_sizes_and_dst_info.data()) + buf_sizes_size);
+/**
+ * @brief Create an instance of `packed_src_and_dst_pointers` populating destination
+ * partitition buffers (if any) from `out_buffers`. In the chunked_pack case
+ * `out_buffers` is empty, and the destination pointer is provided separately
+ * to the `copy_partitions` kernel.
+ *
+ * @param input source table view
+ * @param num_partitions the number of partitions create (1 meaning no splits)
+ * @param num_src_bufs number of buffers for the source columns including children
+ * @param out_buffers the destination buffers per partition if in the non-chunked case
+ * @param stream Optional CUDA stream on which to execute kernels
+ * @param mr RMM memory resource
+ *
+ * @returns new unique pointer to packed_src_and_dst_pointers
+ */
+std::unique_ptr<packed_src_and_dst_pointers> setup_src_and_dst_pointers(
+  cudf::table_view const& input,
+  std::size_t num_partitions,
+  cudf::size_type num_src_bufs,
+  std::vector<rmm::device_buffer>& out_buffers,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  auto src_and_dst_pointers =
```

Comment:
`const` whenever possible. This will be const pointer to non-const data.
```suggestion
  auto const src_and_dst_pointers =
```

Reply:
I tried to make this change @ttnghia, but I am getting build failures. Because I am returning `src_and_dst_pointers`, making it const here looks to want to invoke the copy constructor (which is deleted for `unique_ptr`)

Any ideas?